src/devices/machine/sis5513_ide.cpp

```diff
@@ -0,0 +1,280 @@
+// license: BSD-3-Clause
+// copyright-holders: Angelo Salese
+/**************************************************************************************************
+
+	SiS 5513 IDE controller
+
+    TODO:
+	- Stub interface, to be improved;
+
+**************************************************************************************************/
+
+#include "emu.h"
+#include "sis5513_ide.h"
+
+#define LOG_IO     (1U << 1) // log PCI register accesses
+#define LOG_TODO   (1U << 2) // log unimplemented registers
+#define LOG_MAP    (1U << 3) // log full remaps
+
+#define VERBOSE (LOG_GENERAL | LOG_IO | LOG_TODO | LOG_MAP)
+//#define LOG_OUTPUT_FUNC osd_printf_warning
+
+#include "logmacro.h"
+
+#define LOGIO(...)     LOGMASKED(LOG_IO,   __VA_ARGS__)
+#define LOGMAP(...)    LOGMASKED(LOG_MAP,  __VA_ARGS__)
+#define LOGTODO(...)   LOGMASKED(LOG_TODO, __VA_ARGS__)
+
+DEFINE_DEVICE_TYPE(SIS5513_IDE, sis5513_ide_device, "sis5513_ide", "SiS 5513 IDE Controller")
+
+sis5513_ide_device::sis5513_ide_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: pci_device(mconfig, SIS5513_IDE, tag, owner, clock)
+	, m_ide1(*this, "ide1")
+	, m_ide2(*this, "ide2")
+	, m_irq_pri_callback(*this)
+	, m_irq_sec_callback(*this)
+{
+	// IDE controller with 0xd0 as programming i/f "ATA Host Adapters standard"
+	// pclass bits 1-3 are actually 1 when controller is in native mode
+	// pclass bits 0-2 can be r/w from $09
+	set_ids(0x10395513, 0xd0, 0x010180, 0x00);
+}
+
+void sis5513_ide_device::device_add_mconfig(machine_config &config)
+{
+	BUS_MASTER_IDE_CONTROLLER(config, m_ide1).options(ata_devices, "hdd", nullptr, false);
+	m_ide1->irq_handler().set([this](int state) { m_irq_pri_callback(state); });
+	m_ide1->set_bus_master_space(":maincpu", AS_PROGRAM);
+
+	BUS_MASTER_IDE_CONTROLLER(config, m_ide2).options(ata_devices, "cdrom", nullptr, false);
+	m_ide2->irq_handler().set([this](int state) { m_irq_sec_callback(state); });
+	m_ide2->set_bus_master_space(":maincpu", AS_PROGRAM);
```

Comment:
How is this supposed to work when the host CPU isn’t `:maincpu`?  I realise the i82371 does the same thing, but it looks really fragile and unlikely to work in more exotic setups.